src/bc_protocol/media_packet.rs

```diff
@@ -0,0 +1,381 @@
+                let check_bytes = &data[8..10];
+                const AAC_VALID: &[u8] = &[0xff, 0xf1];
+                match check_bytes {
+                    AAC_VALID => true,
+                    _ => {
+                        trace!(
+                            "AAC failed header checks: {:x?}",
+                            &data[0..min(MAX_HEADER_SIZE, data.len())]
+                        );
+                        false
+                    }
+                }
+            }
+            MediaDataKind::AudioDataAdpcm => {
+                let check_bytes = &data[8..10];
+                const ADPCM_VALID: &[u8] = &[0x0, 0x1];
+                match check_bytes {
+                    ADPCM_VALID => true,
+                    _ => {
+                        trace!(
+                            "ADPCM failed header checks: {:x?}",
+                            &data[0..min(MAX_HEADER_SIZE, data.len())]
+                        );
+                        false
+                    }
+                }
+            }
+            MediaDataKind::InfoData => {
+                // Not sure how to check this yet. Theres only one per stream at the start though
+                true
+            }
+            MediaDataKind::Unknown | MediaDataKind::Continue => true,
+            MediaDataKind::Invalid => false,
+        }
+    }
+}
+
+pub struct MediaDataSubscriber<'a> {
+    binary_buffer: VecDeque<u8>,
+    bc_sub: &'a BcSubscription<'a>,
+}
+
+impl<'a> MediaDataSubscriber<'a> {
+    pub fn from_bc_sub<'b>(bc_sub: &'b BcSubscription) -> MediaDataSubscriber<'b> {
+        MediaDataSubscriber {
+            binary_buffer: VecDeque::new(),
+            bc_sub: bc_sub,
+        }
+    }
+
+    fn fill_binary_buffer(&mut self, rx_timeout: Duration) -> Result<()> {
+        // Loop messages until we get binary add that data and return
+        loop {
+            let msg = self.bc_sub.rx.recv_timeout(rx_timeout)?;
+            if let BcBody::ModernMsg(ModernMsg {
+                binary: Some(binary),
+                ..
+            }) = msg.body
+            {
+                // Add the new binary to the buffer and return
+                self.binary_buffer.extend(binary);
+                break;
+            }
+        }
+        Ok(())
+    }
+
+    fn advance_to_media_packet(&mut self, rx_timeout: Duration) -> Result<()> {
+        // In the event we get an unknown packet we advance by brute force
+        // reading of bytes to the next valid magic
+        while self.binary_buffer.len() < MAX_HEADER_SIZE {
+            self.fill_binary_buffer(rx_timeout)?;
+        }
+
+        // Check the kind, if its invalid use pop a byte and try again
+        let mut magic =
+            MediaDataSubscriber::get_first_n_deque(&self.binary_buffer, MAX_HEADER_SIZE);
+        if INVALID_MEDIA_PACKETS.contains(&MediaData::kind_from_raw(&magic)) {
+            trace!("Advancing to next know packet header: {:x?}", &magic);
+        }
```

Comment:
I don't think you need this `if` block - it does the same check as the `while` beneath it.

Reply:
This if check is to print once only at the beginning of the loop not on every step of the loop.

Reply:
I'd like to keep it in as its a good way to detect new magic and see when the code is skipping a packet. But if you still want to remove that sort of thing from the trace let me know and I'll axe it.

Reply:
We can leave it but I reserve the right to refactor at some point if I change my mind ;)